include/dali/core/exec/tasking/task.h

```diff
@@ -0,0 +1,522 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_EXEC_TASKING_TASK_H_
+#define DALI_CORE_EXEC_TASKING_TASK_H_
+
+#include <any>
+#include <cassert>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+
+#include "dali/core/exec/tasking/sync.h"
+#include "dali/core/small_vector.h"
+
+namespace dali::tasking {
+
+namespace detail {
+class TaskList;
+
+template <typename T, typename U = void>
+struct is_iterable : std::false_type {};
+
+template <typename T>
+struct is_iterable<T, std::void_t<
+  decltype(*std::begin(std::declval<T>())),
+  decltype(*std::end(std::declval<T>()))>>
+: std::true_type {};
+
+template <typename T>
+constexpr bool is_iterable_v = is_iterable<T>::value;
+
+template <typename T>
+struct is_tuple : std::false_type {};
+
+template <typename... Args>
+struct is_tuple<std::tuple<Args...>> : std::true_type {};
+
+template <typename T>
+constexpr bool is_tuple_v = is_tuple<T>::value;
+
+}  // namespace detail
+
+class Scheduler;
+
+
+class TaskResult {
+ public:
+  template <typename F>
+  void SetResultOf(F &&f) {
+    try {
+      if constexpr (std::is_void_v<decltype(f())>) {
+        f();
+        value_ = void_result();
+      } else {
+        value_ = f();
+      }
+    } catch (...) {
+      exception_ = std::current_exception();
+    }
+  }
+
+  template <typename T>
+  void Set(T &&t) {
+    value_ = std::forward<T>(t);
+  }
+
+  void SetException(std::exception_ptr e) {
+    exception_ = std::move(e);
+  }
+
+  bool Empty() {
+    return !value_.has_value();
+  }
+
+  const std::any &Value() const {
+    if (exception_)
+      std::rethrow_exception(exception_);
+    return value_;
+  }
+
+  template <typename T>
+  T Value() const {
+    if (exception_)
+      std::rethrow_exception(exception_);
+    if constexpr (std::is_void_v<T>) {
+      (void)std::any_cast<void_result>(value_);
```

Comment:
Shouldn't we return that? Now it seems to be no-op.

Reply:
There's a bug here, but we don't need to return anything - the return type is `void` anyway. What's missing is `else`.

Reply:
Well, there's a side effect - if the value is _not_ `void_result`, then this function will throw `bad_any_cast` and we want this to happen.